samples: boards: nrf: system_off: sample cleanup

gmarull opened the pull request.

gmarull committed:
samples: boards: nrf: system_off: s/printk/printf

Use printf instead of printk, samples should prefer standard C APIs.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
samples: boards: nrf: system_off: fix includes

Include only what is needed, also sort includes properly.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
samples: boards: nrf: system_off: remove redundant prj.conf comment

Not needed, already mentioned in sample.yaml.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
samples: boards: nrf: system_off: add missing dependency

APP_RETENTION depends on having CRC enabled.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
samples: boards: nrf: system_off: use standard GPIO API

Wake-up from pin is already supported when configuring GPIO interrupt as
GPIO_INT_LEVEL_ACTIVE, so there's no need to resort to HAL.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented on samples/boards/nrf/system_off/src/main.c:
```diff
@@ -4,29 +4,22 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#include "retained.h"
```
@nordicjm could that sample be moved to the retention API?

nordicjm replied:
It could be, and validity would be inbuilt, though would need overlays for each board

gmarull commented:
@anangl ping

anangl commented on samples/boards/nrf/system_off/prj.conf:
```diff
@@ -1,5 +1,4 @@
 # Required to disable default behavior of deep sleep on timeout
-CONFIG_PM_DEVICE=y
```
This should stay. The purpose of this sample is not only to demonstrate System OFF mode:
https://github.com/zephyrproject-rtos/zephyr/blob/0c55710040747d5984a2594a3061c988519d5ac3/samples/boards/nrf/system_off/README.rst#L9-L18
Moreover, the UART needs to be suspended before entering System OFF (the sample lacks that currently), otherwise the power consumption is above 200 uA instead of a few uA (before 37b26f40a6f671499c192a1ac30e32b633b3378f, the UART was suspended by `pm_state_force`, now it needs to be done explicitly).

gmarull replied:
so we need to find an alternative to the `pm_device_action_run`, it is wrong to use that API directly.

gmarull replied:
this is actually yet another case of misunderstanding what CONFIG_PM_DEVICE is about, ref https://github.com/zephyrproject-rtos/zephyr/pull/62160

gmarull replied:
@anangl would it be sufficient to add `disable-rx` to the UART in an overlay?

anangl replied:
> @anangl would it be sufficient to add `disable-rx` to the UART in an overlay?

I'll check.

anangl replied:
I've checked this. Unfortunately, disabling RX makes difference only in the "Sleep 2 s" state (the current consumption is noticeably lower), but not in System OFF. It must be the enabled UART, not just its receiver, that causes these ~217 uA to be drawn.

gmarull replied:
Ok, I'll leave the pm_device_action_run and try to find an alternative in the near future.

gmarull committed:
samples: boards: nrf: system_off: suspend console

To achieve minimal power consumption, only the console (ie UART) needs
to be suspended. Note that this is a temporary workaround, UART driver
should be using device runtime PM so that this is not needed.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

anangl commented on samples/boards/nrf/system_off/src/main.c:
```diff
@@ -55,22 +52,6 @@ int main(void)
-	printf("Sleep %u s with UART off\n", SLEEP_S);
-	rc = pm_device_action_run(cons, PM_DEVICE_ACTION_SUSPEND);
-	k_sleep(K_SECONDS(SLEEP_S));
-	rc = pm_device_action_run(cons, PM_DEVICE_ACTION_RESUME);
```
[README.rst](https://github.com/zephyrproject-rtos/zephyr/blob/d254faa95ae12e7c8270563ff68762ad7eedb638/samples/boards/nrf/system_off/README.rst) needs to be updated accordingly.

gmarull committed:
samples: boards: nrf: system_off: update readme

- Remove obsolete information
- Remove redundant details, e.g. how to flash
- Adjust output

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>
